### PR TITLE
Feat/narwhals backend

### PR DESCRIPTION
**Pull request by philipp-jung**
## What it does

  Replaces the pandas dependency with the backend‑agnostic narwhals (https://github.com/narwhals-dev/narwhals). This means:

  - we can drop pandas from runtime dependencies entirely
  - tab-err will support pandas and polars (https://github.com/pola-rs/polars) going forward.
  - We have the option for supporting duckdb (https://github.com/duckdb/duckdb), PyArrow, cuDF, and Modin going forward. Support for now is untested and experimental.

I believe this is a big step forward for the project, as it will allow more users to use tab-err.

## How it works

narwhals is a wrapper around the APIs for common DataFrame libraries. This means that we can only rely on the minimal shared subset of capabilities between these libraries - which are enough to do what tab-err needs to achieve.

Dtype parity between libraries is handled by relying on narwhals dtypes and explicitly casting mutated columns back to their original dtype after applying errors. That's not enough to make sure there's equal behavior between polars and pandas, so I added tests to ensure this:

- test_backend_parity.py tests that error scenarios applied with polars and pandas lead to the same corrupted DataFrame
- test_dtype_preservation.py tests that edge-cases (datetime, missing values, categorical values) are properly handled across pandas and polars.

We’ll need additional backend‑specific tests (e.g., DuckDB, PyArrow, cuDF, Modin) to validate feature parity beyond pandas/polars.

### CODE REVIEW

**Review by chandlerNick — CHANGES_REQUESTED** [review]
I just got through all the files in the PR, great work! I do still however have a few comments and concerns I think should be addressed before approval. I will detail them below:

- I think it would be worth re-running the notebooks in `tab_err/examples` with the modified code base. This can give a good idea of whether or not anything broke outside of the testing suite (I did not make the tests I wrote exhaustive, per instructions). I am also planning to do some minor changes as far as __str__ / __repr__ methods, high_level api changes, and rewriting the examples notebooks a little after this PR is done.
- Technically, I don't see any errors but I have a few minor comments
1. In `mid_level.py` line 85, I don't think there is XOR there but was slightly confused by the comment.
2. In `category_swap.py` lines 36, 37, it would make sense to factor out `len(unique_vals)` as I saw you do in another file.
3. In `mistype.py` the `_apply` function raised some questions to Sebastian and I last year, and I still have the same questions. Basically, there is no justification/explanation of the heuristic used to map current to target dtypes and it could be beneficial to do so (as is done elsewhere when heuristics are used).
4. In `permutation.py` line 121, there is a typo in a variable name that should be ..._permutation not ..._permutatin.
5. In `replace.py` line 22, this one wasn't you but since there are changes in the file it could make sense to change the error. Masically the message should say "replace" instead of "permutate."
- Nice job factoring out lambda functions, when there are none, the repo can be parallelized more easily in python.
- Finally, and this one is probably worth discussing with Sebastian since he voiced some concerns last week, but there are some higher-level design questions he was asking of which I will list a few:
1. What happens if Narwhals loses traction and development slows behind that of pandas/polars?
2. Who asked for these changes? We need users.
3. Is this primarily a theoretical gain or is there a practical reason for it?
4. (This one is my question mainly) Do you have applications in mind that require the features introduced in this PR?

I hope this helps and nice work.

**Comment by philipp-jung** [comment]
Hey, thanks for the review @chandlerNick, great feedback -- thanks for grinding through this PR which is definitely a bit too large 😬 

I'll leave my comments on the source code itself as I work through your feedback.

File: examples/Error_Types.ipynb
Comment (philipp-jung) [inline examples/Error_Types.ipynb:1]: I re-ran the thing, as you recommended.

File: tab_err/api/mid_level.py [hunk 1]
```diff
@@ -72,10 +78,20 @@ def create_errors(data: pd.DataFrame, config: MidLevelConfig | dict) -> tuple[pd
             error_type = error_model.error_type
             error_rate = error_model.error_rate
 
-            old_error_mask = error_mask.copy()
-            error_mask = error_mechanism.sample(data, column, error_rate, error_mask)
+            old_error_mask = error_mask.clone()
+            error_mask = error_mechanism.sample(data_nw, column, error_rate, error_mask)
+
+            # Compute new error positions (where mask changed from False to True)
```
Comment (philipp-jung) [inline tab_err/api/mid_level.py:84]: The comment was misleading. I believe it referenced line 89, which is not an XOR, but an AND NOT. I decided to remove the comment as it doesn't add any value.

File: tab_err/error_type/_category_swap.py [hunk 2]
```diff
@@ -26,57 +26,81 @@ def _check_type(data: pd.DataFrame, column: int | str) -> None:
...
+        if series.dtype != nw.Categorical:
+            msg = f"Column {column} does not contain values of the Categorical dtype. Cannot insert Mislabels.\n"
+            msg += "Try casting the column to Categorical dtype."
             raise TypeError(msg)
 
-        if len(series.cat.categories) <= 1:
-            msg = f"Column {column} contains {len(series.cat.categories)} categories. Require at least 2 categories to insert mislabels."
+        # Get unique values to check number of categories
+        unique_vals = series.unique()
+        if len(unique_vals) <= 1:
```
Comment (philipp-jung) [inline tab_err/error_type/_category_swap.py:36]: I cannot follow the remark on this @chandlerNick  🙈 I don't see what it is that I could factor out. Do you mind explaining in more detail?

Comment (chandlerNick) [reply tab_err/error_type/_category_swap.py:36]: Ah, weird I must have misread it. I think you did factor it out already.

**Comment by chandlerNick** [comment]
Alright, with me everything is good. I think you should chat with @se-jaeger about the larger scale concerns (ideally here) before I unblock merging.

File: tab_err/error_type/_mistype.py [hunk 3]
```diff
@@ -2,78 +2,100 @@
...
-        return [col_name for col_name in data.columns.tolist() if data[col_name].dtype != "object"]
-
-    def _apply(self: Mistype, data: pd.DataFrame, error_mask: pd.DataFrame, column: int | str) -> pd.Series:
+        return [col_name for col_name in data.columns if data[col_name].dtype != nw.Object]
+
+    def _convert_dtypes(self, data_arr: np.ndarray, mask_arr: np.ndarray, target_dtype: str) -> np.ndarray:
+        # Convert to object array to allow mixed types
+        result_arr = data_arr.astype(object)
+
+        # Apply type conversion where mask is True
```
Comment (philipp-jung) [inline tab_err/error_type/_mistype.py:35]: I see why this raises questions @se-jaeger @chandlerNick -- it's merely a heuristic, and not even principled as such. When I wrote this, I just hard-coded types that can be converted without raising errors (e.g. int to float).

I'll document this in the code itself. And I'm also open to finding better heuristics, or some principled approach even.

File: tab_err/error_type/_mistype.py [hunk 4]
```diff
@@ -2,78 +2,100 @@
...
 
         if self.config.mistype_dtype is not None:
             if self.config.mistype_dtype not in supported_dtypes:
-                msg = f"Unsupported user-specified dtype {self.config.mistype_dtype}. Supported dtypes as {supported_dtypes}."
+                msg = f"Unsupported user-specified dtype {self.config.mistype_dtype}. Supported dtypes are {supported_dtypes}."
                 raise TypeError(msg)
 
             target_dtype = self.config.mistype_dtype
-        else:  # no user-specified dtype, use heuristict to infer one
+        else:  # no user-specified dtype, use heuristic to infer one
```
Comment (philipp-jung) [inline tab_err/error_type/_mistype.py]: So the idea here is to automatically find a `target_dtype` if the user doesn't define one. It's a fallback that "should just work" out of the box. But not a refined one.

Comment (chandlerNick) [reply tab_err/error_type/_mistype.py]: Ok, perhaps in the future we could look for common mistype errors. Maybe something related to how pandas infers types incorrectly at times. Still kind of a foggy notion at this point.

**Comment by philipp-jung** [comment]
Discussing the PR, I think no consensus was reached in favor for a narwhals backend. We discussed that while narwhals makes the library useful in more settings, coding with it is arguably harder than with pandas.

The PR is stalling and getting outdated as a result. That's why I'm withdrawing the PR. If we ever want to adopt narwhals, the patterns in the PR should still work.
